fastfield_codecs/src/null_index/dense.rs

```diff
@@ -0,0 +1,366 @@
+
+/// For the `DenseCodec`, `data` which contains the encoded blocks.
+/// Each block consists of [u8; 12]. The first 8 bytes is a bitvec for 64 elements.
+/// The last 4 bytes are the offset, the number of set bits so far.
+///
+/// When translating the original index to a dense index, the correct block can be computed
+/// directly `orig_idx/64`. Inside the block the position is `orig_idx%64`.
+///
+/// When translating a dense index to the original index, we can use the offset to find the correct
+/// block. Direct computation is not possible, but we can employ a linear or binary search.
+pub struct DenseCodec {
+    // data consists of blocks of 64 bits.
+    //
+    // The format is &[(u64, u32)]
+    // u64 is the bitvec
+    // u32 is the offset of the block, the number of set bits so far.
+    //
+    // At the end one block is appended, to store the number of values in the index in offset.
+    data: OwnedBytes,
+}
+const ELEMENTS_PER_BLOCK: u32 = 64;
+const BLOCK_BITVEC_SIZE: usize = 8;
+const BLOCK_OFFSET_SIZE: usize = 4;
+const SERIALIZED_BLOCK_SIZE: usize = BLOCK_BITVEC_SIZE + BLOCK_OFFSET_SIZE;
+
+fn count_ones(block: u64, pos_in_block: u32) -> u32 {
+    if pos_in_block == 63 {
+        block.count_ones()
+    } else {
+        let mask = (1u64 << (pos_in_block + 1)) - 1;
+        let masked_block = block & mask;
+        masked_block.count_ones()
+    }
+}
+
+impl DenseCodec {
+    /// Open the DenseCodec from OwnedBytes
+    pub fn open(data: OwnedBytes) -> Self {
+        Self { data }
+    }
+    #[inline]
+    /// Check if value at position is not null.
+    pub fn exists(&self, idx: u32) -> bool {
+        let block_pos = idx / ELEMENTS_PER_BLOCK;
+        let bitvec: u64 = self.block(block_pos as usize);
+
+        let pos_in_block = idx % ELEMENTS_PER_BLOCK;
+
+        get_bit_at(bitvec, pos_in_block)
+    }
+    #[inline]
+    pub(crate) fn block(&self, block_pos: usize) -> u64 {
+        let data = &mut &self.data[block_pos as usize * SERIALIZED_BLOCK_SIZE..];
+
+        let block: u64 =
+            BinarySerializable::deserialize(data).expect("could not read block in null index");
+        block
+    }
+
+    #[inline]
+    /// Returns (bitvec, offset)
+    ///
+    /// offset is the start offset of actual docids in the block.
+    pub(crate) fn block_and_offset(&self, block_pos: u32) -> (u64, u32) {
+        let data = &mut &self.data[block_pos as usize * SERIALIZED_BLOCK_SIZE..];
+
+        let block: u64 =
+            BinarySerializable::deserialize(data).expect("could not read block in null index");
+        let offset: u32 =
+            BinarySerializable::deserialize(data).expect("could not read block in null index");
+        (block, offset)
+    }
+
+    /// Return the number of non-null values in an index
+    pub fn num_non_null_vals(&self) -> u32 {
+        let last_block = (self.data.len() / SERIALIZED_BLOCK_SIZE) - 1;
+        self.block_and_offset(last_block as u32).1
```

Comment:
a struct with the pair would be nicer... see above.

Reply:
I'm not sure if there is a bug or not in that code. Do you add an extra last empty block at the end to make that correct?

Reply:
Yes, I did add the last block, it's also explained in a comment